### PR TITLE
Allow for systemd path used by SuSE variants

### LINKED ISSUE #103: SystemD path in SuSE variants is /usr/lib/systemd, not /lib/systemd
In order to accommodate this, a change is required.

### PR DESCRIPTION
**Pull request by untergeek**
Which is /usr/lib/systemd instead of /lib/systemd

Apologies for the empty line/space corrections picked up by my editor (Atom).

fixes #103

### CODE REVIEW

File: lib/pleaserun/detector.rb [hunk 1]
```diff
@@ -25,14 +25,19 @@ def detect_platform
 
   def detect_systemd
     # Expect a certain directory
-    return false unless File.directory?("/lib/systemd/system")
+    return false unless File.directory?("/lib/systemd/system") || File.directory?("/usr/lib/systemd/system")
+
+    # Verify presence of systemctl before proceeding.  This catches an edge
+    # case in ubuntu 12.04 where the systemctl path exists, but no binary
+    out, status = execute([ "which", "systemctl" ])
```
Comment (jordansissel) [inline lib/pleaserun/detector.rb]: Not all systems have `which` available. Thoughts on having `execute` check for this - I'm thinking scanning ENV["PATH"] manually for the program and returning nonzero status if not found.

Comment (untergeek) [inline lib/pleaserun/detector.rb]: I'm fine with that.  It just needs to return non-zero in the event it's not found. for the original code here to work.

File: lib/pleaserun/detector.rb [hunk 2]
```diff
@@ -25,14 +25,19 @@ def detect_platform
 
   def detect_systemd
     # Expect a certain directory
-    return false unless File.directory?("/lib/systemd/system")
+    return false unless File.directory?("/lib/systemd/system") || File.directory?("/usr/lib/systemd/system")
```
Comment (jordansissel) [inline lib/pleaserun/detector.rb]: I think we can simplify by just checking for /usr/lib/systemd/system, forgetting about /lib/systemd/system. Thoughts?

Comment (untergeek) [inline lib/pleaserun/detector.rb]: That, combined with a test for the `systemctl` binary should suffice.

File: lib/pleaserun/detector.rb (hunk 1 above)
Comment (untergeek) [inline lib/pleaserun/detector.rb]: Do you want me to pull this then, and wait for `execute` to be updated?

File: lib/pleaserun/detector.rb (hunk 2 above)
Comment (richarson) [inline lib/pleaserun/detector.rb]: I just checked and Ubuntu 16.04 doesn't seem to have /usr/lib/systemd/system but /lib/systemd/system.
Wasn't systemd supposed to remove this kind of "diversity" between distros? :-/

Comment (richarson) [inline lib/pleaserun/detector.rb]: CentOS does have /usr/lib/systemd/system and /lib is a symlink to usr/lib, so either check would suffice.
I assume RHEL and Fedora do the same symlinking.

File: lib/pleaserun/detector.rb (hunk 1 above)
Comment (richarson) [inline lib/pleaserun/detector.rb]: Shouldn't the next check catch the existence of the binary?
Quote:
```ruby
# Check the version. If `systemctl` fails, systemd isn't available.
out, status = execute([ "systemctl", "--version" ])
```

Comment (untergeek) [inline lib/pleaserun/detector.rb]: I thought so too, but it didn't.

File: lib/pleaserun/detector.rb (hunk 2 above)
Comment (untergeek) [inline lib/pleaserun/detector.rb]: 16.04 does have `/usr/lib/systemd`:
```
root@ubuntu-xenial:~# ls /usr/lib/systemd/
boot  catalog  network  user  user-generators
```
That's why the change is now going to just use `/usr/lib/systemd`.  Each of the systems had that directory, even if it wasn't where the unit files went.

File: lib/pleaserun/detector.rb (hunk 1 above)
Comment (untergeek) [inline lib/pleaserun/detector.rb]: This is the output that lead to this PR:
```
IOError: Cannot run program "systemctl" (in directory "/root"): error=2, No such file or directory
           popen3 at org/jruby/RubyIO.java:4417
           popen3 at /usr/share/logstash/vendor/jruby/lib/ruby/1.9/open3.rb:74
          execute at /usr/share/logstash/vendor/bundle/jruby/1.9/gems/pleaserun-0.0.21/lib/pleaserun/detector.rb:76
   detect_systemd at /usr/share/logstash/vendor/bundle/jruby/1.9/gems/pleaserun-0.0.21/lib/pleaserun/detector.rb:31
  detect_platform at /usr/share/logstash/vendor/bundle/jruby/1.9/gems/pleaserun-0.0.21/lib/pleaserun/detector.rb:23
           detect at /usr/share/logstash/vendor/bundle/jruby/1.9/gems/pleaserun-0.0.21/lib/pleaserun/detector.rb:17
```

**Comment by jordansissel** [comment]
Pause for a tableflip.

* SuSE uses /usr/lib/systemd and /lib is not a symlink to /usr/lib
* Debian 8 uses /lib/systemd and /lib is not a symlink to /usr/lib
* Fedora 23 uses /usr/lib/systemd and /lib *is* a symlink to /usr/lib.
* Debian 8 has both /lib/systemd and /usr/lib/systemd.

what is this.. I don't even .... 


Ok so, I tried to dig a bit more to see if maybe we can grep our way to success and hoped I could find in a config file where systemd was configured to look for system services. I found nothing so far.

SuSE, Debian 8, and Fedora 23 are all using different paths (sorta...) for the systemd/system/ directory.. Uuugh.

**Comment by jordansissel** [comment]
THE PLOT THICKENS.

systemd apparently can look multiple places. The manpage for `systemd.unit` has this to say:

```
       Table 1.  Load path when running in system mode (--system).
       ┌────────────────────────┬─────────────────────┐
       │Path                    │ Description         │
       ├────────────────────────┼─────────────────────┤
       │/etc/systemd/system     │ Local configuration │
       ├────────────────────────┼─────────────────────┤
       │/run/systemd/system     │ Runtime units       │
       ├────────────────────────┼─────────────────────┤
       │/usr/lib/systemd/system │ Units of installed  │
       │                        │ packages            │
       └────────────────────────┴─────────────────────┘
```

Of course, the exact paths will depend on the linux distro and who compiled systemd with whatever settings. For example, Debian 8's `systemd.unit` doesn't mention `/usr/lib/systemd/system` but has `/lib/systemd/system` listed instead.

**Comment by jordansissel** [comment]
Possibly a light at the end of the tunnel, I accidentally found `systemctl show`:

```
% PAGER=cat systemctl show --property UnitPath
UnitPath=/etc/systemd/system /run/systemd/system /run/systemd/generator /usr/local/lib/systemd/system /usr/lib/systemd/system /run/systemd/generator.late
```

**Comment by jordansissel** [comment]
It seems like /etc/systemd/system is a place where lots of symlinks end up, so.. uhh. Here's a fun hack to find the most likely place we should put unit files:

```
% ls -l /etc/systemd/system/ | grep -- '->' | sed -e 's/^.*> //'  | xargs dirname | sort -n | uniq -c
```

On Fedora 23, the most common is /usr/lib/systemd/system (correct!). On Debian 8, the most common (and only!) for me is /lib/systemd/system (correct!).

**Comment by jordansissel** [comment]
I don't know what the right choice is, yet. Given /etc/systemd/system is present on *all* distros tested so far, it makes sense to maybe have pleaserun target that path. Bonus points that the manpage says that directory is for "local sysadmin stuff" or whatever...

**Comment by jordansissel** [comment]
Another fun attempt - ask systemd for the path ("FragmentPath") of all the service units and pick the most common:

```
systemctl list-unit-files --type service | awk '/\.service/ {print $1}' | xargs -n1 systemctl show --property FragmentPath 2> /dev/null | awk -F= '/FragmentPath/ { print $2 }' | xargs -n1 dirname | sort | uniq -c | sort -rn | head -1 | awk '{$1=""; print}'
```

* Fedora 23: /usr/lib/systemd/system
* Debian 8.4:  /lib/systemd/system
* CentOS 7: /usr/lib/systemd/system
* CoreOS 899.17.0: /usr/lib64/systemd/system
* Ubuntu 16.04: /lib/systemd/system

Welp. So much for hardcoding this :)

**Comment by jordansissel** [comment]
Given tonight's experiments and docs review, I think I am happy with having pleaserun default to writing systemd units to `/etc/systemd/system` -- @untergeek thoughts? Can you check this on SuSE?

**Comment by nadrimajstor** [comment]
For what it's worth, as a noobish linux sysadmin, I could easily wrap my brain around idea: if a conf file comes from a package it goes in to some upstream/vendor path and just works with sensible defaults - /etc is left for me to provide changes to those defaults.

**Comment by untergeek** [comment]
I will test it on my SuSE vm and see what shakes out.

**Comment by untergeek** [comment]
The template has to be updated to launch the prerestart script `/etc/systemd/system`, but it does in fact work.  My concern, looking at what's _in_ the `/etc/systemd/system` directory is that it contains the unit files for the services that need to start earliest: dbus, getty, multi-user, network, sockets, etc.

Will Logstash be started late enough to have all dependencies at startup time?

**Comment by jordansissel** [comment]
Systemd lets you specify dependencies, so we can have it start after
whatever we need. Probably after networking or multi-user

File: lib/pleaserun/detector.rb (hunk 1 above)
Comment (richarson) [inline lib/pleaserun/detector.rb]: Ah, OK then

**Comment by jordansissel** [comment]
Some friends on twitter pointed me at this cool thing:

```
pkg-config systemd --variable=systemdsystemunitdir
```

Tested on CentOS 7, Fedora 23, and Debian 8, and it correctly identifies the systemd/system/ location.

**Comment by jordansissel** [comment]
Except sometimes `pkg-config` isn't installed even if systemd is used. But I did find that (On openSUSE's installer, for example) systemd ships /usr/share/pkgconfig/systemd.pc even if `pkg-config` is not present, so maybe we can .... probably all hope is lost ... 

Maybe two tries for guessing systemd's path:
* try the `pkg-config` thing
* if that fails, try the `systemctl list-unit-files --type service | awk '/\.service/ {print $1}' | xargs -n1 systemctl show --property FragmentPath 2> /dev/null | awk -F= '/FragmentPath/ { print $2 }' | xargs -n1 dirname | sort | uniq -c | sort -rn | head -1 | awk '{$1=""; print}'` method.

**Comment by jordansissel** [comment]
commit 1223773 makes a new attribute for systemd `unit_path` that defaults to `/etc/systemd/system` based on the research from this issue.

**Comment by untergeek** [comment]
is this issue even necessary after that patch, then?

**Comment by untergeek** [comment]
It seems the better test in `detect_systemd` is the existence of the `systemd` binary, rather than an unpredictable path.  What do you think, @jordansissel ?

**Comment by jordansissel** [comment]
Agreed, maybe we just check for systemctl and call it a day, since the path is not constant.

**Comment by jordansissel** [comment]
@untergeek yeah I think I accidentally made this PR unnecessary while I was trying to find a solution to the systemd problem. Onward! <3

**Comment by untergeek** [comment]
Closing.  Made irrelevant by https://github.com/jordansissel/pleaserun/commit/122377358bcfd8188dba0b672af8fbe538a66515
